src/ray/common/ray_syncer/ray_syncer.h

```diff
@@ -0,0 +1,212 @@
+  /// for the connection.
+  ///
+  /// \param connection The connection to the remote node.
+  void Connect(std::unique_ptr<NodeSyncConnection> connection);
+
+  /// Connect to a node.
+  /// TODO (iycheng): Introduce grpc channel pool and use node_id
+  /// for the connection.
+  ///
+  /// \param connection The connection to the remote node.
+  void Connect(std::shared_ptr<grpc::Channel> channel);
+
+  void Disconnect(const std::string &node_id);
+
+  /// Register the components to the syncer module. Syncer will make sure eventually
+  /// it'll have a global view of the cluster.
+  ///
+  /// \param component_id The component to sync.
+  /// \param reporter The local component to be broadcasted.
+  /// \param receiver The snapshot of the component in the cluster.
+  /// \param pull_from_reporter_interval_ms The frequence to pull a message from reporter
+  /// and push it to sending queue.
+  bool Register(RayComponentId component_id,
+                const ReporterInterface *reporter,
+                ReceiverInterface *receiver,
+                int64_t pull_from_reporter_interval_ms = 100);
+
+  /// Function to broadcast the messages to other nodes.
+  /// A message will be sent to a node if that node doesn't have this message.
+  /// The message can be generated by local reporter or received by the other node.
+  ///
+  /// \param message The message to be broadcasted.
+  void BroadcastMessage(std::shared_ptr<const RaySyncMessage> message);
+
+  /// Get the current node id.
+  const std::string &GetNodeId() const { return node_id_; }
+
+  /// Get the io_context used by RaySyncer.
+  instrumented_io_context &GetIOContext() { return io_context_; }
+
+  NodeSyncConnection *GetSyncConnection(const std::string &node_id) {
+    auto iter = sync_connections_.find(node_id);
+    if (iter == sync_connections_.end()) {
+      return nullptr;
+    }
+    return iter->second.get();
+  }
+
+ private:
+  /// io_context for this thread
+  instrumented_io_context &io_context_;
+
+  /// The current node id.
+  const std::string node_id_;
+
+  /// Manage connections
+  absl::flat_hash_map<std::string, std::unique_ptr<NodeSyncConnection>> sync_connections_;
+
+  /// The local node state
+  std::unique_ptr<NodeState> node_state_;
+
+  /// If the field is set to be true, it means, message generated or received
+  /// is needed to be sent to other nodes.
+  Array<bool> component_broadcast_;
+
+  /// Timer is used to do broadcasting.
+  ray::PeriodicalRunner timer_;
+
+  FRIEND_TEST(SyncerTest, Test1To1);
+};
+
+class ClientSyncConnection;
+class ServerSyncConnection;
+
+/// RaySyncerService is a service to take care of resource synchronization
+/// related operations.
+/// Right now only raylet needs to setup this service. But in the future,
+/// we can use this to construct more complicated resource reporting algorithm,
+/// like tree-based one.
+class RaySyncerService : public ray::rpc::syncer::RaySyncer::CallbackService {
```

Comment:
Does this mean we are going to use a different gRPC path from the existing one?

Reply:
Yes.

Reply:
I had the same concern as https://github.com/ray-project/ray/pull/23270#issuecomment-1075807762. I think the plan sounds good to me. (I think alternative we can do is to replace the existing completion queue impl to use callback APIs?)

Reply:
@rkooo567 I remember there is a PR (https://github.com/ray-project/ray/pull/15279) doing this. Let's get more idea about this API and we should be more confident to deprecate the old one later.